python/examples/atomistic/3-atomistic-model-with-nl.py

```diff
@@ -0,0 +1,300 @@
+# <https://github.com/Luthaf/metatensor-lj-test/blob/main/src/metatensor_lj_test/pure.py>`_.
+# The linked version also implements ``per_atom`` energies as well as atom selection
+# using the ``selected_atoms`` parameter of the ``forward()`` method. In this model, we
+# shift the energy by its value at the ``cutoff``. This will break the conservativeness
+# of the potential, which is unproblematic in here because we use a large cutoff and
+# therefore the potential already almost decayed to zero. For more sophisticated methods
+# like a polynomial switching potential like XPLOR, we refer to the literature.
+
+
+class LennardJonesModel(torch.nn.Module):
+    """Implementation of a single particle type Lennard-Jones potential."""
+
+    def __init__(self, cutoff, sigma, epsilon):
+        super().__init__()
+        self._nl_options = NeighborListOptions(cutoff=cutoff, full_list=False)
+
+        self._sigma = sigma
+        self._epsilon = epsilon
+
+        # shift the energy to 0 at the cutoff
+        self._shift = 4 * epsilon * ((sigma / cutoff) ** 12 - (sigma / cutoff) ** 6)
+
+    def requested_neighbor_lists(self) -> List[NeighborListOptions]:
+        return [self._nl_options]
+
+    def forward(
+        self,
+        systems: List[System],
+        outputs: Dict[str, ModelOutput],
+        selected_atoms: Optional[Labels],
+    ) -> Dict[str, TensorMap]:
+        if list(outputs.keys()) != ["energy"]:
+            raise ValueError(
+                "'energy' is the only supported `outputs` key; found "
+                f"{', '.join(outputs.keys())}"
+            )
+
+        # we don't want to worry about selected_atoms yet
+        if selected_atoms is not None:
+            raise NotImplementedError("selected_atoms is not implemented")
+
+        if outputs["energy"].per_atom:
+            raise NotImplementedError("per atom energy is not implemented")
+
+        all_energies = []
+
+        # Initialize device so we can access it outside of the for loop
+        device = torch.device("cpu")
+        for system in systems:
+            device = system.device
+            dtype = system.positions.dtype
+
+            neighbors = system.get_neighbor_list(self._nl_options)
+            all_i = neighbors.samples.column("first_atom").to(torch.long)
+            all_j = neighbors.samples.column("second_atom").to(torch.long)
+
+            energy = torch.zeros(len(system), dtype=dtype, device=device)
+
+            distances = neighbors.values.reshape(-1, 3)
+
+            sigma_r_6 = (self._sigma / torch.linalg.vector_norm(distances, dim=1)) ** 6
+            sigma_r_12 = sigma_r_6 * sigma_r_6
+            e = 4.0 * self._epsilon * (sigma_r_12 - sigma_r_6) - self._shift
+
+            # We only compute each pair once (full_list=False in self._nl_options),
+            # and assign half of the energy to each atom
+            energy = energy.index_add(0, all_i, e, alpha=0.5)
+            energy = energy.index_add(0, all_j, e, alpha=0.5)
```

Comment:
I don't think we need this if we are not doing `per_atom`. We can directly sum `e` and store that.

Reply:
Or we keep the `per_atom` option and use it to explain the data in the samples?

Reply:
I think we can leave it out. It is not important for ase. But it is for lammps, right?

Reply:
no, it's fine (the user choses to request per-atom energies or not). What's important for LAMMPS is selected_atoms since we always have to subselect the real atoms from the ghosts.